### ISSUE TITLE
Game State

### ISSUE THREAD

**Issue by weiffert** [weiffert/Component-Based-Game-Engine#5]
Just put everything referring to game states here. Anything is welcome.

**Comment by weiffert** [comment]
Changes:
StateManager is now BaseState.
State variables are protected, not private.
changeState functions are in game.
getNumber() and getId() are no longer pure virtuals, but defined in BaseState.
id and number are no longer constant.

**Comment by weiffert** [comment]
Ok. Process of the game. The game is opened. Time passes for initializing welcome screen. Welcome screen. time passes to initialize loading screen. Loading Screen and loading of all the other states. Then Menu. Selection of game. Game. selection of pause. Pause. Then back to game. Game. The level ends. End. Then the kill screen ends and Menu. There is no time lag between switching the states of menu, level, pause, and end because they are already initialized with the loading state. The initialized states have to be stored somewhere. If the game is played again, the level state has to reset. So there would be a loading screen during initializing (update function for loading). Then the game is closed and the states are destroyed. I think this is the ideal flowchart of time and states.

Basically, the states have to be pre-initialized and stored somewhere.

**Comment by weiffert** [comment]
For pre-initializing the menu, level, pause, and end states, it happens during loading(). A vector of *BaseState in systemManager could hold these states. Implementation could be in the loading update function like if _____State has not been initialized, initialize it once per render. That is easily testable with the vector. If id != "StateWelcome" or any other state, initialize and push back. For states that change, delete the state after they have been used to reset them. Then, initialize them with loading for a level.

For changing states, the game pointer could be set to a pointer from the state vector in systemManager with the proper id or number. This would be a pre-initialized state. This allows for there to be no loading time between states.

**Comment by weiffert** [comment]
Does this make sense? Do you think it will work?

**Comment by weiffert** [comment]
The issue is now initializing entities. How to do this is the issue. It clearly happens in loading, but then, how does anything have the proper entities? Welcome should be moved after loading, so that it has the proper entities. This makes it a lot easier to initialize entities. I don't know how to do this yet and still have the loading screen update.

Thoughts?
